components/domino-logic/src/main/kotlin/net/corda/lifecycle/domino/logic/BlockingDominoTile.kt

```
@@ -0,0 +1,89 @@
+import net.corda.lifecycle.LifecycleEventHandler
+import net.corda.lifecycle.LifecycleStatus
+import net.corda.lifecycle.StopEvent
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * This class encapsulates domino logic for components that have no dependencies, where we want to delay startup until we receive an
+ * external signal.
+ *
+ * @param startTile complete this future to signal that this tile should start. This is a one shot processes i.e. if the Tile starts,
+ * then stops and then starts again, there is no way to delay the Tile starting for the second time.
+ */
+class BlockingDominoTile(componentName: String,
+                         coordinatorFactory: LifecycleCoordinatorFactory,
+                         private val startTile: CompletableFuture<Unit>
+): DominoTile {
+    companion object {
+        private val instancesIndex = ConcurrentHashMap<String, Int>()
+    }
+
+    override val coordinatorName: LifecycleCoordinatorName by lazy {
+        LifecycleCoordinatorName(
+            componentName + "Blocking",
+            instancesIndex.compute(this::class.java.simpleName) { _, last ->
+                if (last == null) {
+                    1
+                } else {
+                    last + 1
+                }
+            }.toString()
+        )
+    }
+
+    override val coordinator = coordinatorFactory.createCoordinator(coordinatorName, EventHandler())
+    private val internalState = AtomicReference(LifecycleStatus.DOWN)
+
+    override val dependentChildren: Collection<LifecycleCoordinatorName> = emptyList()
+
+    override fun start() {
+        coordinator.start()
+        startTile.whenComplete { _, exception ->
+            if (exception == null) {
+                coordinator.postEvent(AsynchronousReady)
+            } else {
+                coordinator.postEvent(AsynchronousException(exception))
```

I don't think this can recover from an error...

---

Yes, do you think that is a problem? This can only happen if the future gets completed exceptionally, not sure how we should recover in that situation. Without making the interface more complex.

---

yes, seems like there's no scenario where the future can fail, right. Maybe once we have one, we might have more information on how to adjust this to recover from it.

---

Yes I we don't complete the future exceptionally anywhere in our code base at the moment.

---

So, maybe we should ignore the exception and add a comment in the documentation?

---

I think as the Tile is designed with a one shot future it makes the most sense to keep the current behaviour. i.e. if we get an error in the future then never start the tile. I will add a line in the documentation.

Currently the future can only be completed once so the if the Tile restarts the future will already be completed. There is no way to delay the tile starting for the second time. If the future completes exceptionally the current behaviour seems sensible to me i.e. the Tile never starts.

---

I added a comment to the doc describing current behaviour.